python_modules/dagster/dagster/_core/definitions/module_loaders/object_list.py

```diff
@@ -145,9 +145,19 @@ def _do_collision_detection(self) -> None:
         for key, asset_objects in self.asset_objects_by_key.items():
             # In certain cases, we allow asset specs to collide because we know we aren't loading them.
             # If there is more than one asset_object in the list for a given key, and the objects do not refer to the same asset_object in memory, we have a collision.
-            num_distinct_objects_for_key = len(
-                set(id(asset_object) for asset_object in asset_objects)
-            )
+
+            # special handling for conflicting AssetSpecs since they are value objects and we can compare them
```

Comment:
are automation conditions value objects?

Reply:
I'm also eyeing PartitionsDefinitions and AssetDeps (that contain PartitionMappings)

Reply:
Ya the `==` will recurse down through the tuples occasionally hitting plain classes and doing is same instance checks which I think is fine. Automation conditions are all (or almost all) `@record` so `PartitionsDefinition` / `PartitionMapping` are the known in memory instance compares which is par for the course for them. 

The motivating report for this was programatically generated `AssetSpec` with i believe just `key` so the idea here is at least handling simple dupes is helpful.

Reply:
updated comment and tests in response, let me know if that helps